language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java

```diff
@@ -510,4 +514,30 @@ public void didClose(DidCloseTextDocumentParams params) {
     @Override
     public void didSave(DidSaveTextDocumentParams params) {
     }
+
+    @JsonRequest
+    public CompletableFuture<List<FoldingRange>> foldingRange(FoldingRangeRequestParams params) {
+        return CompletableFuture.supplyAsync(() -> {
+            List<FoldingRange> emptyFoldingRangeList = new ArrayList<>();
+            String fileUri = params.getTextDocument().getUri();
+            Optional<Path> filePath = CommonUtil.getPathFromURI(fileUri);
+            if (filePath.isEmpty() || CommonUtil.isCachedExternalSource(fileUri)) {
+                return emptyFoldingRangeList;
+            }
+            try {
+                Optional<Document> document = workspaceManager.document(filePath.get());
```

Comment:
Without accessing the document at this point, let's create a context and pass the context to `FoldingRangeProvider.getFoldingRange` API. then all the calculations can be carry out within the API

Reply:
Updated.